k2/python/csrc/fsa.cc

```diff
@@ -6,192 +6,73 @@
-      std::ostringstream os;
-      os << self;
-      return os.str();
-    });
-
-py::class_<Fsa>(m, "Fsa")
-    .def(py::init<>())
-    .def("num_states", &Fsa::NumStates)
-    .def("final_state", &Fsa::FinalState)
-    .def("__str__", [](const Fsa &self) { return FsaToString(self); })
-    .def_readwrite("arc_indexes", &Fsa::arc_indexes)
-    .def_readwrite("arcs", &Fsa::arcs);
-
-py::class_<std::vector<Fsa>>(m, "FsaVec")
-    .def(py::init<>())
-    .def("clear", &std::vector<Fsa>::clear)
-    .def("__len__", [](const std::vector<Fsa> &self) { return self.size(); })
-    .def("push_back",
-         [](std::vector<Fsa> *self, const Fsa &fsa) { self->push_back(fsa); })
-    .def("__iter__",
-         [](const std::vector<Fsa> &self) {
-           return py::make_iterator(self.begin(), self.end());
-         },
-         py::keep_alive<0, 1>());
-// py::keep_alive<Nurse, Patient>
-// 0 is the return value and 1 is the first argument.
-// Keep the patient (i.e., `self`) alive as long as the Nurse (i.e., the
-// return value) is not freed.
-
-py::class_<std::vector<Arc>>(m, "ArcVec")
-    .def(py::init<>())
-    .def("clear", &std::vector<Arc>::clear)
-    .def("__len__", [](const std::vector<Arc> &self) { return self.size(); })
-    .def("__iter__",
-         [](const std::vector<Arc> &self) {
-           return py::make_iterator(self.begin(), self.end());
-         },
-         py::keep_alive<0, 1>());
-
-py::class_<Cfsa>(m, "Cfsa")
-    .def(py::init<>())
-    .def(py::init<const Fsa &>(), py::arg("fsa"), py::keep_alive<1, 2>())
-    .def("num_states", &Cfsa::NumStates)
-    .def("num_arcs", &Cfsa::NumArcs)
-    .def("arc",
-         [](Cfsa *self, int s) {
-           DCHECK_GE(s, 0);
-           DCHECK_LT(s, self->NumStates());
-           auto begin = self->arc_indexes[s];
-           auto end = self->arc_indexes[s + 1];
-           return py::make_iterator(self->arcs + begin, self->arcs + end);
-         },
-         py::keep_alive<0, 1>())
-    .def("__str__",
-         [](const Cfsa &self) {
-           std::ostringstream os;
-           os << self;
-           return os.str();
-         })
-    .def("__eq__",  // for test only
-         [](const Cfsa &self, const Cfsa &other) { return self == other; });
-
-py::class_<std::vector<Cfsa>>(m, "CfsaStdVec")
-    .def(py::init<>())
-    .def("clear", &std::vector<Cfsa>::clear)
-    .def("push_back", [](std::vector<Cfsa> *self,
-                         const Cfsa &cfsa) { self->push_back(cfsa); })
-    .def("__len__", [](const std::vector<Cfsa> &self) { return self.size(); })
-    .def("__iter__",
-         [](const std::vector<Cfsa> &self) {
-           return py::make_iterator(self.begin(), self.end());
-         },
-         py::keep_alive<0, 1>());
-  PybindCfsaVec(m);
-*/
+  // Note(fangjun): Users are not supposed to use `k2::Fsa` directly
+  // in Python; the following wrapper is only used by pybind11 internally
+  // so that it knows `k2::_Fsa` is a subclass of `k2::Fsa`.
```

Comment:
2 underscores or 1?

Reply:
`k2::_Fsa` is used only in this `.cc` file and is not available to other files.
I could not find out a good name, so I just use `class _Fsa`. Any name except `class Fsa` is fine here.

Reply:
For python users, we have

```python
import k2
assert '__Fsa' not in dir(k2)
assert 'Fsa' in dir(k2)
````